internal/component/snapshotlease/snapshotlease.go

```diff
@@ -0,0 +1,202 @@
+	// ErrDeleteSnapshotLease indicates an error in deleting the snapshot lease resources.
+	ErrDeleteSnapshotLease druidv1alpha1.ErrorCode = "ERR_DELETE_SNAPSHOT_LEASE"
+)
+
+type _resource struct {
+	client client.Client
+}
+
+// New returns a new snapshot lease component operator.
+func New(client client.Client) component.Operator {
+	return &_resource{
+		client: client,
+	}
+}
+
+// GetExistingResourceNames returns the names of the existing snapshot leases for the given Etcd.
+func (r _resource) GetExistingResourceNames(ctx component.OperatorContext, etcdObjMeta metav1.ObjectMeta) ([]string, error) {
+	resourceNames := make([]string, 0, 2)
+	// We have to get snapshot leases one lease at a time and cannot use label-selector based listing
+	// because currently snapshot lease do not have proper labels on them. In this new code
+	// we will add the labels.
+	// TODO: Once all snapshot leases have a purpose label on them, then we can use List instead of individual Get calls.
+	deltaSnapshotObjectKey := client.ObjectKey{Name: druidv1alpha1.GetDeltaSnapshotLeaseName(etcdObjMeta), Namespace: etcdObjMeta.Namespace}
+	deltaSnapshotLease, err := r.getLeasePartialObjectMetadata(ctx, deltaSnapshotObjectKey)
+	if err != nil {
+		return resourceNames, &druiderr.DruidError{
+			Code:      ErrGetSnapshotLease,
+			Cause:     err,
+			Operation: "GetExistingResourceNames",
+			Message:   fmt.Sprintf("Error getting delta snapshot lease: %v for etcd: %v", deltaSnapshotObjectKey, druidv1alpha1.GetNamespaceName(etcdObjMeta)),
+		}
+	}
+	if deltaSnapshotLease != nil && metav1.IsControlledBy(deltaSnapshotLease, &etcdObjMeta) {
+		resourceNames = append(resourceNames, deltaSnapshotLease.Name)
+	}
+	fullSnapshotObjectKey := client.ObjectKey{Name: druidv1alpha1.GetFullSnapshotLeaseName(etcdObjMeta), Namespace: etcdObjMeta.Namespace}
+	fullSnapshotLease, err := r.getLeasePartialObjectMetadata(ctx, fullSnapshotObjectKey)
+	if err != nil {
+		return resourceNames, &druiderr.DruidError{
+			Code:      ErrGetSnapshotLease,
+			Cause:     err,
+			Operation: "GetExistingResourceNames",
+			Message:   fmt.Sprintf("Error getting full snapshot lease: %v for etcd: %v", fullSnapshotObjectKey, druidv1alpha1.GetNamespaceName(etcdObjMeta)),
+		}
+	}
+	if fullSnapshotLease != nil && metav1.IsControlledBy(fullSnapshotLease, &etcdObjMeta) {
+		resourceNames = append(resourceNames, fullSnapshotLease.Name)
+	}
+	return resourceNames, nil
+}
+
+// Sync creates or updates the snapshot leases for the given Etcd.
+// If backups are disabled for the Etcd resource, any existing snapshot leases are deleted.
+func (r _resource) Sync(ctx component.OperatorContext, etcd *druidv1alpha1.Etcd) error {
+	if !etcd.IsBackupStoreEnabled() {
+		ctx.Logger.Info("Backup has been disabled. Triggering deletion of snapshot leases")
+		return r.deleteAllSnapshotLeases(ctx, etcd.ObjectMeta, func(err error) error {
+			return druiderr.WrapError(err,
```

Comment:
Same as above. `DeleteAllOf` is different from `Delete`, since deletion of a single resource can return a `NotFound` error, but not a `DeleteAllOf` call.